kernel/kernel-api/src/main/java/io/delta/kernel/internal/util/SchemaUtils.java

```diff
@@ -79,6 +83,83 @@ public static void validateSchema(StructType schema, boolean isColumnMappingEnab
     validateSupportedType(schema);
   }
 
+  /**
+   * Performs the following validations on an updated table schema:
+   *
+   * <p>The following checks are performed:
+   *
+   * <ul>
+   *   <li>No duplicate columns are allowed
+   *   <li>Names contain only valid characters
+   *   <li>Data types are supported
+   *   <li>No new non-nullable fields are added
+   *   <li>No type promotions are performed
+   *   <li>Physical column name consistency is preserved in the new schema
+   *   <li>Partition columns are not modified
+   *   <li>Nested IDs for array/map types are preserved in the new schema
+   * </ul>
+   */
+  public static void validateUpdatedSchema(
+      StructType currentSchema,
+      StructType newSchema,
+      List<String> partitionCols,
+      Metadata metadata) {
+    validateSchema(newSchema, true /*columnMappingEnabled*/);
+    ColumnMapping.validateColumnIds(
+        currentSchema, newSchema, TableConfig.ICEBERG_COMPAT_V2_ENABLED.fromMetadata(metadata));
+    validateUpdatedSchemaCompatibility(currentSchema, newSchema, partitionCols);
+  }
+
+  /* Verifies that no non-nullable fields are added, no type promotions are performed and that
+  partition columns are not modified */
+  private static void validateUpdatedSchemaCompatibility(
+      StructType currentSchema, StructType newSchema, List<String> partitionCols) {
+    validatePartitionColumns(newSchema, partitionCols);
+    // Identify added columns based on field IDs
+    Map<Integer, StructField> newSchemaIdToField = idToField(newSchema);
```

Comment:
How do we validate that the new schema doesn't have any duplicate field ids

Reply:
That's validated above, as we recurse we're keeping track of already seen field IDs and if we encounter something in the set we'll throw an exception

https://github.com/delta-io/delta/pull/4196/files#diff-38491854ca9c7c7860acd852eef7e8477196d0c9690ea55ed1691ada0c8bf631R242